### PR TITLE
feat(fs): gave nodes identity and timestamps, added rename and symbolic links

### PR DESCRIPTION
**Pull request by FlareCoding**
## Summary
- Every file reported inode 0 and time 0, so compilers deduplicating include paths collapsed them into one and build tools could not tell what had changed, rename and symbolic links did not exist, and the initrd loader silently dropped the links it contained.
- Nodes carry a lifetime-stable identity and access, modification, and change times, directories share one implementation, and rename, symbolic links, O_TRUNC, and O_EXCL behave as programs expect.
- Path resolution follows links at constant stack depth, and scratch buffers moved off the privileged heap in line with the dynamic privilege model.

---

> [!NOTE]
> **Medium Risk**
> Large VFS surface area (symlink resolution, rename locking, open/create edge cases) can cause subtle path or concurrency bugs; changes are heavily tested in `fs.test.cpp` but touch core namespace behavior used by all file syscalls.
> 
> **Overview**
> **VFS nodes now expose real metadata and POSIX-style namespace operations** so userland build tools and libc expectations (stat, rename, symlinks) work on ramfs/devfs.
> 
> Nodes get **per-instance `dev` IDs**, **stable `ino` values**, and **nanosecond atime/mtime/ctime** (via new `clock::realtime_ns()`). Base `node::getattr`/`setattr` and `fsetattr` replace duplicated device-node `getattr` overrides. **Shared `fs::dir_node`** centralizes lookup/readdir, bounded teardown, and **rename/rmdir** with a global tree lock and ordered directory locking to avoid deadlocks.
> 
> **Path resolution** follows symlinks (heap-spliced paths, `SYMLOOP_MAX`), supports **`LOOKUP_NOFOLLOW`**, and **`open` with `O_CREAT`** can create through dangling symlink chains. **`rename`**, **`symlink`**, **`O_TRUNC`** (writable regular files only), and **`O_EXCL`** are wired through VFS and Linux **`rename`/`renameat`/`symlink`/`symlinkat`** syscalls; **stat/getdents** report **ino/dev/timestamps**. **CPIO initrd** extraction now creates **symlinks** and restores **mtime** from the archive.
> 
> **Clock** drops per-arch cached boot time; **`boot_realtime_ns()`** and **`realtime_ns()`** are inlined off **`rtc::boot_unix_ns()`**. Syscall path scratch buffers move from **`kzalloc` to `uzalloc`** where appropriate.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 6dbcf7adbc2c105e0adfde5db84c5a2352cd629a. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: kernel/fs/fs.cpp [hunk 1]
```diff
@@ -633,19 +788,30 @@ file* open_at(node* base_dir, const char* path, uint32_t flags, int32_t* out_err
...
                 err = parent->lookup(name, name_len, &n);
-                if (err == ERR_NOENT) {
+                if (err == OK && (flags & O_EXCL)) {
+                    release_node_ref(n);
+                    n = nullptr;
+                    err = ERR_EXIST;
+                } else if (err == OK && n->type() == node_type::symlink) {
+                    release_node_ref(n);
+                    n = nullptr;
+                    err = resolve_path_at_internal(base_dir, path, true, &n);
```
[the remark this answers is not in the document]
Comment (FlareCoding) [reply kernel/fs/fs.cpp]: fixed

File: kernel/fs/ramfs/ramfs.cpp [hunk 2]
```diff
@@ -250,60 +177,91 @@ int32_t dir_node::rmdir(const char* name, size_t len) {
         return fs::ERR_NOTDIR;
     }
 
-    auto* child_dir = static_cast<dir_node*>(child);
-    if (child_dir->m_child_count > 0) {
-        return fs::ERR_NOTEMPTY;
+    return remove_empty_dir(static_cast<fs::dir_node*>(child));
```
[the remark this answers is not in the document]
Comment (FlareCoding) [reply kernel/fs/ramfs/ramfs.cpp]: fixed

File: kernel/fs/dir_node.cpp [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+        return ERR_NOTDIR;
+    }
+
+    if (!existing_is_dir) {
+        detach_child(existing);
+        return OK;
+    }
+
+    return remove_empty_dir(static_cast<dir_node*>(existing));
+}
```
[the remark this answers is not in the document]
Comment (FlareCoding) [reply kernel/fs/dir_node.cpp:205]: fixed

File: kernel/fs/fs.cpp [hunk 4]
```diff
@@ -592,6 +747,69 @@ __PRIVILEGED_CODE int32_t unmount(const char* target) {
...
+        if (err != OK) {
+            break;
+        }
+
+        release_node_ref(link);
+        link = found;
+        if (link->type() != node_type::symlink) {
+            link->add_ref();
+            *out = link;
+            break;
```
[the remark this answers is not in the document]
Comment (FlareCoding) [reply kernel/fs/fs.cpp:794]: fixed

File: kernel/fs/fs.cpp [hunk 5]
```diff
@@ -592,6 +747,69 @@ __PRIVILEGED_CODE int32_t unmount(const char* target) {
...
+        }
+
+        release_node_ref(dir);
+        dir = parent;
+
+        node* found = nullptr;
+        err = dir->lookup(name, name_len, &found);
+        if (err == ERR_NOENT) {
+            err = dir->create(name, name_len, 0, out);
+            break;
```
[the remark this answers is not in the document]
Comment (FlareCoding) [reply kernel/fs/fs.cpp:785]: fixed
